### ISSUE TITLE
[19.11] secure variable set on an environment is not properly persisted

### ISSUE THREAD

**Issue by k-c-p** [gocd/gocd#7481]
##### Issue Type

- Bug Report

##### Summary

A secure environment variable set on an environment is not properly persistent and later returned as empty.

Worked pretty fine up to and including version 19.10, but seems to be broken in 19.11.

###### Basic environment details

* Go Version: 19.11
* JAVA Version: 13.0.1
* OS: Debian Jessie

##### Steps to Reproduce

Create an environment and set a secure variable `ENV_SEC_VAR` on the environment. Then create a pipeline which executes `/bin/sh -c printenv` to dump all environment variables.

##### Expected Results
The output should yield

```
ENV_SEC_VAR=******
```

##### Actual Results
But instead I get:

```
ENV_SEC_VAR=
```

##### Possible Fix

I colleague of mine dug into the communication between the frontend and the Go Server. He found that this is sent

```json
{
    "environment_variables": {
        "add": [
            {
                "name": "broken_env",
                "value": "",
                "encrypted_value": "broken stuff",
                "secure": true
            }
        ],
        "remove": []
    }
}
```
but it should be this

```json
{
    "environment_variables": {
        "add": [
            {
                "name": "secret_env",
                "value": "secret stuff",
                "secure": true
            }
        ],
        "remove": []
    }
}
```

##### Log snippets

logs do not show any errors.

**Comment by GaneshSPatil** [comment]
Hi @k-c-p, Thanks for reporting the issue.
I can confirm that this issue is reproducible on GoCD release `v19.11.0`.

#### Impact of the Issue:
The above-mentioned bug causes the secure environment variables to be saved with an empty value on the server while editing the environment from the new environments SPA.
This can cause pipelines to fail, as the required environment variables coming from the environment are not available at the time of pipeline execution.

#### Scope of the Issue:
The above-mentioned bug is only applicable to the environments created/modified via the new environments page on GoCD v19.11.0. 
All the existing environments created prior to the `19.11.0` upgrade are unaffected.

#### Cause:
The new environments SPA was released as part of the GoCD release `v19.11.0`.  While saving an environment from this new page, invalid data was sent to the server causing the empty value to be set for the secure environment variables.

#### Workaround for users on GoCD release `v19.11.0`

##### 1. Use Old Environments Page.
Users can toggle back to the old environments page by setting the `show_new_environments_spa` toggle value to `Off`. Use the following command to turn off the toggle:

```console
curl 'https://go-server-url:8154/go/api/admin/feature_toggles/show_new_environments_spa' -H 'Accept: application/vnd.go.cd.v1+json' -X PUT -d '{"toggle_value": "Off"}' -u 'username:password'
```

##### 2. Use [Environment Config API](https://api.gocd.org/current/#environment-config) to modify environment(s).


#### Fix:
A [pull request](https://github.com/gocd/gocd/pull/7485) has been raised to fix this issue which is planned for the upcoming GoCD release `v19.12.0` (due next week). Please subscribe to the pull request for any further updates on this issue.

**Comment by rajiesh** [comment]
Verified on 19.12.0 (10888-29b0f854605987c8edab9dced4814b62dc751a11)
